Feature/72 73 : Api endpoint to serve data and authentication with data validation

ArmaanKatyal opened the pull request:
In this pull request, I have: 
- A Functioning REST endpoint to get, post, delete and update the data. Create models for the endpoint to include the student's name (string), email (string), ICard number (string), active status (bool), ISAF-paying status (bool)
- Created routes for the endpoint
- Created controllers for the endpoint
- Created necessary services
- added middleware such that the app shouldn't be able to make API requests unless they have the proper authentication

Extra:
- I also added data validation using JOI, thought that it was necessary to validate data before making amends to DB
- Added tests for all the routes; I can add more unit tests. let me know if that is necessary

ArmaanKatyal committed:
Feature #62: Create an express API that connects to MongoDB

ArmaanKatyal committed:
Feature #62: Create an express API that connects to MongoDB

ArmaanKatyal committed:
set the node and commonjs to true

ArmaanKatyal committed:
updates

ArmaanKatyal committed:
updates

ArmaanKatyal committed:
set the port number to default

ArmaanKatyal committed:
added setTimeout to avoid test failure as the MongoDB connection is slow

ArmaanKatyal committed:
added readme with short desc on how to run server and test

ArmaanKatyal committed:
added jest to eslint config

ArmaanKatyal committed:
updates

ArmaanKatyal committed:
changed the port number

ArmaanKatyal committed:
updates

ArmaanKatyal committed:
added routes for students and auth

ArmaanKatyal committed:
code for issues: 72,73,78

ArmaanKatyal committed:
updates

ArmaanKatyal committed:
added cors support

ArmaanKatyal committed:
update

ArmaanKatyal committed:
fixed the port issue for heroku

ArmaanKatyal committed:
Procfile for heroku

ArmaanKatyal committed:
update

ArmaanKatyal committed:
added the controller to register student

ArmaanKatyal committed:
fixed issue where student was not updating

ArmaanKatyal committed:
fixed issue where __v was inserted in db

ArmaanKatyal committed:
fixed issue where __v was inserted in db

ArmaanKatyal committed:
added route for register controller

ArmaanKatyal committed:
added validation for register req.body

ArmaanKatyal committed:
removed deployment files

ArmaanKatyal committed:
updated the readme

ArmaanKatyal committed:
fixed db conn issue during testing

ArmaanKatyal committed:
added user details in response when login request is successful

ArmaanKatyal committed:
added status code and fixed update response issue

ArmaanKatyal committed:
updates

ArmaanKatyal committed:
fixed the tests to run sequentially

ArmaanKatyal committed:
Merge branch 'develop' into Feature/72-73

dimassudjito commented on backend/src/app.js:
```diff
@@ -2,23 +2,43 @@ const express = require('express')
-    useNewUrlParser: true,
-    useUnifiedTopology: true,
-})
+if (process.env.NODE_ENV === 'test') {
```
can we change 'test' into 'stag' for staging (the other option which we can also explicitly define in the else statement is 'prod' for production)

dimassudjito replied:
we should also change the database name in MongoDB appropriately

ArmaanKatyal replied:
Okay, so I will make the appropriate changes to API endpoints so that only the request with the required API key in the header can make the request. That will be stored privately in the app, I guess ?. 

Regarding line 13 in src/app.js -> I can't change the test option to stag because JEST automatically sets NODE_ENV='test' whenever tests are executed. I need that to test API endpoints on the test database rather than the actual database. If there is a workaround, I will look into it and make the changes if possible.

Thank you for reviewing the work and replying so quickly. 👍🏽

ArmaanKatyal replied:
> we should also change the database name in MongoDB appropriately

Yes, we can change the database name in MongoDB. Just let me know what should I name them or naming scheme should I follow

dimassudjito replied:
> Okay, so I will make the appropriate changes to API endpoints so that only the request with the required API key in the header can make the request. That will be stored privately in the app, I guess ?.

Yes, it will be used from the environment variable

dimassudjito replied:
> Regarding line 13 in src/app.js -> I can't change the test option to stag because JEST automatically sets NODE_ENV='test' whenever tests are executed. I need that to test API endpoints on the test database rather than the actual database. If there is a workaround, I will look into it and make the changes if possible.

I see, that's a misunderstanding on my side; I didn't think "test" is for automated tests. When we deploy the backend, we  want to have 2 version: staging and production. The staging one will be used for our development, while the production one for actual use by users. We can set different MongoDB URI similar to how we differentiate for testing

dimassudjito replied:
staging will be connected to our develop branch while production will be connected to our main branch

ArmaanKatyal replied:
Sure we can do that.

ArmaanKatyal replied:
I have added the code block that will connect the backend to different databases based on the environment variable version provided to it.

dimassudjito reviewed (changes requested):
This is looking really good! I'm impressed by your pace of work. One big thing is how we're guarding the API endpoint. I know I haven't made a detailed description of the issue because I didn't expect anyone to pick it up in this sprint, but we'll only be guarding the API with a generic API-key that the client app (the mobile app, the admin site) needs to have in order to access the API. 

Users in general won't need to login to use the app (e.g., restaurants doesn't need to login to scan ICard) so we won't guard the API with a user authentication. It's very likely that we'll use google oauth to do login for students in the future. That being said, you've built a solid login system and we can definitely keep it for now.

ArmaanKatyal committed:
added the middleware to check the api-key

ArmaanKatyal committed:
added the new middleware

ArmaanKatyal committed:
fixed the testcases to check for api-key and commented the auth test for now

ArmaanKatyal committed:
update

ArmaanKatyal committed:
added feature to connect to different databases based on the development and production

ArmaanKatyal committed:
updates

ArmaanKatyal committed:
Create Procfile

dimassudjito commented on backend/src/app.js:
```diff
@@ -2,23 +2,52 @@ const express = require('express')
+    res.json({ status: db.readyState, database: db.name })
 })
 
+app.use('/api/auth', authRouter) // auth routes
```
we can exclude the auth router from the app.js for now

ArmaanKatyal replied:
yeah, I have removed auth router

dimassudjito commented on backend/src/controllers/Students.js:
```diff
@@ -0,0 +1,64 @@
+const Student = require('../models/Student')
+const validateStudent = require('../validation/Student')
+
```
Could we add a getAll endpoint as well? Will be super useful for future admin site

ArmaanKatyal replied:
okay, I have added the getAll endpoint

dimassudjito commented on backend/src/controllers/Students.js:
```diff
@@ -0,0 +1,64 @@
+            _id: req.params.studentId,
+        })
+        if (!checkStudent)
+            return res.status(400).json({ message: 'Student does not exist' }) // if student does not exist, return 400 with error message
```
(nit) I think your code is descriptive enough that you don't need to comment on this line for example (this is a very good thing). We just need to comment on lines where it's not obvious what's going on.

ArmaanKatyal replied:
okay I will remove all the unnecessary comments

dimassudjito commented on backend/src/controllers/Students.js:
```diff
@@ -0,0 +1,64 @@
+            return res.status(400).json({ message: 'Student does not exist' }) // if student does not exist, return 400 with error message
+
+        try {
+            // Delete the student
```
IMO it's okay to keep this kind of comment because it's like a title to the code block (just giving example on what to comment and not)

ArmaanKatyal replied:
yeah I have removed this comment

dimassudjito commented on backend/src/models/Student.js:
```diff
@@ -0,0 +1,36 @@
+
+const StudentSchema = new mongoose.Schema(
+    {
+        // create a new schema for the student
```
(nit) unnecessary comment

ArmaanKatyal replied:
okay. Removed

dimassudjito commented on backend/src/models/Student.js:
```diff
@@ -0,0 +1,36 @@
+        },
+    },
+    { collection: 'students', versionKey: false }
+) // set the collection name to temp, this is just for testing
```
is this still a relevant comment? it's no longer named temp right

ArmaanKatyal replied:
oh, I left it there by mistake. I have removed it

dimassudjito commented on backend/src/routes/Students.js:
```diff
@@ -0,0 +1,13 @@
+const router = express.Router()
+
+// verifyToken is a middleware that checks if the user is authenticated and has the correct api-key
+router.get('/:studentId', verifyApiKey, studentController.getStudent) // get a student by id
```
add route for getAll after adding it in controller

ArmaanKatyal replied:
done

dimassudjito commented on backend/src/server.js:
```diff
@@ -1,5 +1,5 @@
 const app = require('./app')
+require('dotenv').config()
+// const port = 8080
```
we can delete this commented line

ArmaanKatyal replied:
deleted the comment

dimassudjito commented on backend/src/authentication/verifyToken.js:
```diff
@@ -0,0 +1,40 @@
+const jwt = require('jsonwebtoken')
```
could we move this file to a src/services folder instead of authentication folder

ArmaanKatyal replied:
sure. I will do that

dimassudjito reviewed (changes requested):
Very happy with the implementation results; JOI does an awesome work at validating the data input. Don't worry about deploying and dockerizing the backend in this PR (you can handle it in another branch/PR). After you resolve my comments we can definitely merge it soon

dimassudjito commented:
**Reviewer's checklist**

General Check

- [ ]  Check that the code compiles and runs
- [ ]  Check that the code is producing the required output by comparing it to the acceptance criteria in the GitHub issues. (e.g., compare it with Figma sketch for frontend work)
- [ ]  Check that the code has acceptable tests and that the tests are passing.
- [ ]  Check that the PR is directed at the correct branch (develop Vs. main)
- [ ]  Check that the PR does not have merge conflict [Dimas]
- [ ]  Check that CI/CD tests are passing

Implementation Review

- [ ]  Understand implementation approach and give feedback on how we can improve the general approach
- [ ]  Understand data structures being used in the implementation and provide feedback (i.e., is it optimal?)
- [ ]  Understand external libraries used in the implementation and provide feedback (i.e., is it deprecated? Is there better libraries out there?)
- [ ]  Check for repeated code and give feedback
- [ ]  Check for performance-intensive code (e.g., heavy algorithm) and give feedback
- [ ]  Check for dead code or commented code that should be deleted
- [ ]  Check if variable names are sensible and descriptive

ArmaanKatyal committed:
removed the auth routes

ArmaanKatyal committed:
updated the folder location

ArmaanKatyal committed:
added new controller and removed unnecessary comments

ArmaanKatyal committed:
removed unnecessary comments

ArmaanKatyal committed:
added new route

ArmaanKatyal committed:
updated the folder name from authentication to services

ArmaanKatyal commented on backend/tests/TestSuite.test.js:
```diff
@@ -0,0 +1,131 @@
+        })
+    })
+
+    // describe('test auth', () => {
+    //     test('POST /api/auth/register -> should return status 201, message as Student Registered', async () => {
+    //         const response = await request(app)
+    //             .post('/api/auth/register')
+    //             .send({
+    //                 name: 'Demo User1',
+    //                 email: 'demouser1@ualberta.ca',
+    //                 password: 'demo@123',
+    //                 icard_number: '123456789',
+    //             })
+    //         expect(response.statusCode).toBe(201)
+    //         expect(response.body.message).toBe('Student Registered')
+    //     })
+
+    //     test('POST /api/auth/login -> should return token with status 200', async () => {
+    //         const response = await request(app).post('/api/auth/login').send({
+    //             email: 'demouser1@ualberta.ca',
+    //             password: 'demo@123',
+    //         })
+    //         expect(response.statusCode).toBe(200) // expect the status code to be 200
+    //     })
+
```
okay I have moved the tests to different files and removed this commented auth test block

ArmaanKatyal commented on backend/tests/TestSuite.test.js:
```diff
@@ -0,0 +1,131 @@
+const request = require('supertest')
```
done

ArmaanKatyal committed:
added flag to run jest sequentially to avoid errors and race condition

ArmaanKatyal committed:
update

ArmaanKatyal committed:
segregated the tests into different files

ArmaanKatyal committed:
segregated the tests into different files

ArmaanKatyal committed:
Delete Procfile

dimassudjito commented:
@ArmaanKatyal So sorry I missed this point yesterday, but could we also add an endpoint to get a student by their ICard number (e.g., getByICardNumber or something like that). This is gonna be very important for the next sprint when we hook the scanner in the mobile app to the backend

ArmaanKatyal committed:
added a new controller to get student by icard

ArmaanKatyal committed:
added a new route

ArmaanKatyal committed:
added new tests

ArmaanKatyal committed:
update

ArmaanKatyal committed:
Merge remote-tracking branch 'origin/Feature/72-73' into Feature/72-73

ArmaanKatyal commented:
> @ArmaanKatyal So sorry I missed this point yesterday, but could we also add an endpoint to get a student by their ICard number (e.g., getByICardNumber or something like that). This is gonna be very important for the next sprint when we hook the scanner in the mobile app to the backend

I have added the new controller and route to get a student by their ICard number and I have also added a test for the same.

dimassudjito reviewed (approved):
Great job on this PR! I have tested all endpoints from postman and I'm happy with the result.
